File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    private int[][] dirs = new int[][]{{0,1},{1,0},{0,-1},{-1,0}};
    public void dfs(int r, int c, int[][] grid, int t, int[][] time){
        for(int[] dir : dirs){
            int newr = r+dir[0];
            int newc = c+dir[1];
            if(newr >= grid.length || newr < 0 || newc >= grid[r].length || newc < 0) continue;
            if(grid[newr][newc] == 1 || (grid[newr][newc] == -1 && time[newr][newc] > t+1)){
                grid[newr][newc] = -1;
                time[newr][newc] = t+1;
                dfs(newr, newc, grid, t+1, time);
            }
        }
    }
    public int orangesRotting(int[][] grid) {
        int m = grid.length, n = grid[0].length, res = 0;;
        if(m == 0 || n == 0) return 0;
        int[][] time = new int[m][n];
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == 2) time[i][j] = 0;
                if(grid[i][j] == 1) time[i][j] = Integer.MAX_VALUE;
            }
        }
        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j < grid[i].length; j++){
                if(grid[i][j] == 2){
                    dfs(i, j, grid, 0, time);
                }
            }
        }
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == 1) return -1;
                else if(grid[i][j] == -1) res = Math.max(res, time[i][j]);
            }
        }
        return res;
    }
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def __init__(self):
        self.dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    def dfs(self, r, c, grid, t, time):
        for dir in self.dirs:
            newr, newc = r + dir[0], c + dir[1]
            if (
                newr >= len(grid)
                or newr < 0
                or newc >= len(grid[r])
                or newc < 0
            ):
                continue
            if (
                grid[newr][newc] == 1
                or (grid[newr][newc] == -1 and time[newr][newc] > t + 1)
            ):
                grid[newr][newc] = -1
                time[newr][newc] = min(time[newr][newc], t + 1)
                self.dfs(newr, newc, grid, t + 1, time)
                
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m, n, res = len(grid), len(grid[0]), 0
        if m == 0 or n == 0:
            return 0

        time = [[0] * n for _ in range(m)]

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    time[i][j] = 0
                if grid[i][j] == 1:
                    time[i][j] = float('inf')

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    self.dfs(i, j, grid, 0, time)

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    return -1
                elif grid[i][j] == -1:
                    res = max(res, time[i][j])

        return res
        